src/samplebuffer.h

```diff
@@ -0,0 +1,90 @@
+#ifndef SAMPLEBUFFER_H
+#define SAMPLEBUFFER_H
+
+#include "util/types.h" // CSAMPLE
+
+#include <algorithm> // std::swap
+
+// A sample buffer with properly aligned memory to enable SSE optimizations.
+//
+// No resize operation is provided intentionally for maximum efficiency!
+// If the size of an existing sample buffer needs to be altered after
+// construction this can simply be achieved by swapping the contents with
+// a temporary sample buffer that has been constructed with the appropriate
+// size:
+//
+//     SampleBuffer sampleBuffer(oldSize);
+//     ...
+//     SampleBuffer(newSize).swap(sampleBuffer);
+//     // Now the data in sampleBuffer is uninitialized and
+//     // sampleBuffer.size() == newSize
+//
+// No sample data is copied when swapping the contents of two samples buffers
+// for resizing!
+class SampleBuffer {
+    Q_DISABLE_COPY(SampleBuffer);
+
+  public:
+    typedef size_t size_type;
+    typedef CSAMPLE value_type;
```

Comment:
I know the arguments pro typedefs and they are valid. 
But they become less important in the age or refactoring tools and intellisense. 
When a user reads size_type, it can be everything int is somehow well defined. 
In Eclipse you will see nice popups showing the function prototype. getting to know what 
size_type will cost an extra click, loosing the current editing context. 
This argument becomes stronger if we have typedefs for a small scopes like here. 

It may rise the question: This buffer returns a value_type, aha. but I need a buffer for CSAMPLE ... 

It is very unlikely that this calls will change its type but CSAMPLE not.

Such small typedef scopes also required type casts. Every typecast is a candidate for an error. I prefer to keep a values under the same type during the whole lifetime. 

On the other hand I support application wide typedefs. And I like you idea to introduce a global type for all CSAMPLE indexes (like we did for gain values)